--- env_vars_util.py	2025-05-12 09:11:38.632398009 -0400
+++ env_temp.py	2025-05-12 09:13:25.925221003 -0400
@@ -182,7 +182,7 @@
         new_vars['#'] = ia_t, str(len(input_args))
         for i, arg in enumerate(input_args):
             index = i + 1
-            new_vars[str(index)] = input_args[i]
+            new_vars[str(index)] = (ia_t, input_args[i])
 
     if not exit_status is None:
         new_vars['?'] = es_t, exit_status
